rules/DeadCode/Rector/Assign/RemoveUnusedVariableAssignRector.php

```diff
@@ -82,20 +88,65 @@ public function refactor(Node $node): ?Node
         return $node;
     }
 
-    /**
-     * @return Variable[]
-     */
-    private function findVariableUsages(FunctionLike $functionLike, Assign $assign): array
+    private function isUsed(Assign $assign, Variable $variable): bool
     {
-        return $this->betterNodeFinder->find((array) $functionLike->getStmts(), function (Node $node) use (
-            $assign
+        $isUsedPrev = (bool) $this->betterNodeFinder->findFirstPreviousOfNode($variable, function (Node $node) use (
+            $variable
         ): bool {
-            if (! $node instanceof Variable) {
-                return false;
-            }
+            return $this->isVariableNamed($node, $variable);
+        });
 
-            // skip assign value
-            return $assign->var !== $node;
+        if ($isUsedPrev) {
+            return true;
+        }
+
+        $isUsedNext = (bool) $this->betterNodeFinder->findFirstNext($variable, function (Node $node) use (
+            $variable
+        ): bool {
+            return $this->isVariableNamed($node, $variable);
         });
+
+        if ($isUsedNext) {
+            return true;
+        }
+
+        /** @var FuncCall|MethodCall|New_|NullsafeMethodCall|StaticCall $expr  */
+        $expr = $assign->expr;
+        if (! $this->isCall($expr)) {
+            return false;
+        }
+
+        $args = $expr->args;
+        foreach ($args as $arg) {
+            $variable = $arg->value;
+            if (! $variable instanceof Variable) {
+                continue;
+            }
+
+            $previousAssign = $this->betterNodeFinder->findFirstPreviousOfNode($assign, function (Node $node) use (
+                $variable
+            ): bool {
+                return $node instanceof Assign && $this->isVariableNamed($node->var, $variable);
+            });
+            if ($previousAssign instanceof Assign) {
+                return $this->isUsed($assign, $variable);
+            }
+        }
+
+        return false;
+    }
+
+    private function isCall(Expr $expr) : bool
+    {
+        return $expr instanceof FuncCall || $expr instanceof MethodCall || $expr instanceof New_ || $expr instanceof NullsafeMethodCall || $expr instanceof StaticCall;
```

Comment:
could this be simplified with FunctionLike?

Reply:
no, these are `Expr`.